### PR TITLE
Add Checked Throws to Stream Serialization APIs

### PR DESCRIPTION
**Pull request by alzimmermsft**
# Description

Adds checked `IOException` to `JsonReader`, `JsonWriter`, and `JsonSerializable` and `XMLStreamException` to `XmlReader`, `XmlWriter`, and `XmlSerializable`. Introduces `Function`-like and `Consumer`-like interfaces that are checked with either `IOException` or `XMLStreamException` to simplify try-catching when using `Function` and `Consumer`-based APIs on the streaming interfaces.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/core/azure-json/src/main/java/com/azure/json/IOExceptionBiConsumer.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+/**
+ * Equivalent in functionality to {@link BiConsumer} except that {@link #accept(Object, Object)} is checked with
+ * {@link IOException}.
+ *
+ * @param <T> Type of the first input.
+ * @param <U> Type of the second input.
+ */
+@FunctionalInterface
+public interface IOExceptionBiConsumer<T, U> {
```
Comment (lmolkova) [inline sdk/core/azure-json/src/main/java/com/azure/json/IOExceptionBiConsumer.java:17]: do you want users to use it?

File: sdk/core/azure-json/src/main/java/com/azure/json/IOExceptionFunction.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+/**
+ * Equivalent in functionality to {@link Function} except that {@link #apply(Object)} is checked with
+ * {@link IOException}.
+ *
+ * @param <T> Input type of the function.
+ * @param <R> Output type of the function.
+ */
+@FunctionalInterface
+public interface IOExceptionFunction<T, R> {
```
Comment (lmolkova) [inline sdk/core/azure-json/src/main/java/com/azure/json/IOExceptionFunction.java:17]: same here, do we really need to expose it?

Comment (alzimmermsft) [reply sdk/core/azure-json/src/main/java/com/azure/json/IOExceptionFunction.java:17]: This is taken in a few `JsonReader` APIs and it greatly simplifies the need to handle try-catching in lambdas. Same goes for the similar interfaces that match Java interfaces but with checked exceptions.

**Review by lmolkova — APPROVED** [review]
LGTM!

**Pull request merged**